README.md
[![Build Status][badge-travis-image]][badge-travis-url]

# kong-plugin-aws-lambda

Invoke an [AWS Lambda](https://aws.amazon.com/lambda/) function from Kong. It can be used in combination with other request plugins to secure, manage or extend the function.

## Configuration

### Enabling the plugin on a Service

#### With a database

Configure this plugin on a [Service](https://docs.konghq.com/latest/admin-api/#service-object) by making the following request:

```
$ curl -X POST http://kong:8001/services/{service}/plugins \
  --data name=aws-lambda \
  --data "config.aws_region=AWS_REGION" \
  --data "config.function_name=LAMBDA_FUNCTION_NAME"
```

#### Without a database

Configure this plugin on a [Service](https://docs.konghq.com/latest/admin-api/#service-object) by adding this section to your declarative configuration file:

```
plugins:
- name: aws-lambda
  service: {service}
  config:
    aws_region: AWS_REGION
    function_name: LAMBDA_FUNCTION_NAME
```

In both cases, `{service}` is the `id` or `name` of the Service that this plugin configuration will target.


### Enabling the plugin on a Route

#### With a database

Configure this plugin on a [Route](https://docs.konghq.com/latest/admin-api/#Route-object) with:

```
$ curl -X POST http://kong:8001/routes/{route}/plugins \
  --data name=aws-lambda \
  --data "config.aws_region=AWS_REGION" \
  --data "config.function_name=LAMBDA_FUNCTION_NAME"
```

#### Without a database

Configure this plugin on a [Route](https://docs.konghq.com/latest/admin-api/#route-object) by adding this section to your declarative configuration file:

```
plugins:
- name: aws-lambda
  route: {route}
  config:
    aws_region: AWS_REGION
    function_name: LAMBDA_FUNCTION_NAME
```

In both cases, `{route}` is the `id` or `name` of the Route that this plugin configuration will target.

### Enabling the plugin on a Consumer

#### With a database

You can use the `http://localhost:8001/plugins` endpoint to enable this plugin on specific [Consumers](https://docs.konghq.com/latest/admin-api/#Consumer-object):

```
$ curl -X POST http://kong:8001/consumers/{consumer}/plugins \
  --data name=aws-lambda \
  --data "config.aws_region=AWS_REGION" \
  --data "config.function_name=LAMBDA_FUNCTION_NAME"
```

#### Without a database

Configure this plugin on a [Consumer](https://docs.konghq.com/latest/admin-api/#Consumer-object) by adding this section to your declarative configuration file:

```
plugins:
- name: aws-lambda
  route: {route}
  config:
    aws_region: AWS_REGION
    function_name: LAMBDA_FUNCTION_NAME
```

In both cases, `{consumer}` is the `id` or `username` of the Consumer that this plugin configuration will target.

You can combine `consumer_id` and `service_id`

In the same request, to furthermore narrow the scope of the plugin.

### Global plugins

- **Using a database**, all plugins can be configured using the `http://kong:8001/plugins/` endpoint.
- **Without a database**, all plugins can be configured via the `plugins:` entry on the declarative configuration file.

A plugin which is not associated to any Service, Route or Consumer (or API, if you are using an older version of Kong) is considered "global", and will be run on every request. Read the [Plugin Reference](https://docs.konghq.com/latest/admin-api/#add-plugin) and the [Plugin Precedence](https://docs.konghq.com/latest/admin-api/#precedence)sections for more information.

## Parameters

Here's a list of all the parameters which can be used in this plugin's configuration:

| Form Parameter | default | description
|----------------|---------|-------------
| `name`|| The name of the plugin to use, in this case: `aws-lambda`.
| `service_id`|| The id of the Service which this plugin will target.
| `route_id` || The id of the Route which this plugin will target.
| `enabled` | `true` | Whether this plugin will be applied.
| `consumer_id` || The id of the Consumer which this plugin will target.
|`config.aws_key` <br>*semi-optional* || The AWS key credential to be used when invoking the function. This value is required if `aws_secret` is defined.
|`config.aws_secret` <br>*semi-optional* ||The AWS secret credential to be used when invoking the function. This value is required if `aws_key` is defined.
|`config.aws_region` <br>*semi-optional* || The AWS region where the Lambda function is located. The plugin *does not* attempt to validate the provided region name; an invalid region name will result in a DNS name resolution error. This value cannot be specified if `host` is set.
|`config.aws_role_session_name` <br>*semi-optional* || The session name of the AWS role is used by EKS hosting to assume a role other than the one defined in the `AWS_WEB_IDENTITY_TOKEN_FILE` system variable provided by the service account. This value is used if the value `aws_key` is not defined. It's better to define the `aws_region` to reduce the latency when calling STS.
|`config.aws_assume_role_arn` <br>*semi-optional* || The assume role is used to assume a role different than the role defined by the variable AWS_ROLE_ARN in the EKS environment. This value is used if the value `aws_key` is not defined. This value offers the possibility to use a cross account role.
|`config.function_name` || The AWS Lambda function name to invoke.
|`config.timeout`| `60000` | Timeout protection in milliseconds when invoking the function.
|`config.keepalive`| `60000` | Max idle timeout in milliseconds when invoking the function.
|`config.qualifier` <br>*optional* || The [`Qualifier`](http://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_RequestSyntax) to use when invoking the function.
|`config.invocation_type` <br>*optional*| `RequestResponse` | The [`InvocationType`](http://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_RequestSyntax) to use when invoking the function. Available types are `RequestResponse`, `Event`, `DryRun`.
|`config.log_type` <br>*optional* | `Tail`| The [`LogType`](http://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_RequestSyntax) to use when invoking the function. By default `None` and `Tail` are supported.
|`config.host` <br>*semi-optional* || The AWS lambda host. If not specified, `aws_region` is required and the official AWS lambda endpoint for the given AWS region is used as host.
|`config.port` <br>*optional* | `443` | The TCP port that this plugin will use to connect to the server.
|`config.unhandled_status` <br>*optional* | `200`, `202` or `204` | The response status code to use (instead of the default `200`, `202`, or `204`) in the case of an [`Unhandled` Function Error](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html#API_Invoke_ResponseSyntax)
|`config.forward_request_body` <br>*optional* | `false` | An optional value that defines whether the request body is to be sent in the `request_body` field of the JSON-encoded request. If the body arguments can be parsed, they will be sent in the separate `request_body_args` field of the request. The body arguments can be parsed for `application/json`, `application/x-www-form-urlencoded`, and `multipart/form-data` content types.
|`config.forward_request_headers` <br>*optional* | `false` | An optional value that defines whether the original HTTP request headers are to be sent as a map in the `request_headers` field of the JSON-encoded request.
|`config.forward_request_method` <br>*optional* | `false` | An optional value that defines whether the original HTTP request method verb is to be sent in the `request_method` field of the JSON-encoded request.
|`config.forward_request_uri` <br>*optional* |`false`|An optional value that defines whether the original HTTP request URI is to be sent in the `request_uri` field of the JSON-encoded request. Request URI arguments (if any) will be sent in the separate `request_uri_args` field of the JSON body.
|`config.is_proxy_integration` <br>*optional* | `false` | An optional value that defines whether the response format to receive from the Lambda to [this format](https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format). Note that the parameter `isBase64Encoded` is not implemented.
|`config.awsgateway_compatible` <br>*optional* | `false` | An optional value that defines whether the plugin should wrap requests into the Amazon API gateway.
|`config.proxy_url` <br>*semi-optional* || An optional value that defines whether the plugin should connect through the given proxy server URL. This value is required if `proxy_scheme` is defined.
|`config.proxy_scheme` <br>*semi-optional* || An optional value that defines which HTTP protocol scheme to use in order to connect through the proxy server. The schemes supported are: `http` and `https`. This value is required if `proxy_url` is defined.
|`config.skip_large_bodies` <br>*optional* | `true` | An optional value that defines whether very large bodies (that are buffered to disk) should be sent by Kong. Note that sending very large bodies will have an impact on the system memory.

## Notes

If you do not provide `aws.key` or `aws.secret`, the plugin uses an IAM role inherited from the instance running Kong.

First, the plugin will try ECS metadata to get the role. If no ECS metadata is available, the plugin will fall back on EC2 metadata.

[badge-travis-image]: https://travis-ci.com/Kong/kong-plugin-aws-lambda.svg?branch=master
[badge-travis-url]: https://travis-ci.com/Kong/kong-plugin-aws-lambda


CHANGELOG.md
# Kong AWS Lambda plugin changelog

### Releasing new versions

- update changelog below
- update rockspec version
- update version in `handler.lua`
- commit as `chore(*) release x.y.z`
- tag commit as `x.y.z`
- push commit and tags
- upload to luarocks; `luarocks upload kong-plugin-aws-lambda-x.y.z-1.rockspec --api-key=abc...`
- test rockspec; `luarocks install kong-plugin-aws-lambda`


## aws-lambda 3.6.1 22-Oct-2020

- feat: adding support to use service account role in EKS and also assume a given role

## aws-lambda 3.5.1 6-Okt-2020

- fix: `skip_large_bodies` to honor config setting

## aws-lambda 3.5.0 22-Sep-2020

- feat: adding support for 'isBase64Encoded' flag in Lambda function responses
- fix: respect `skip_large_bodies` config setting even when not using
  AWS API Gateway compatibility

## aws-lambda 3.4.0 12-May-2020

- Change `luaossl` to `lua-resty-openssl`
- fix: do not validate region name against hardcoded list of regions
- feat: add `host` configuration to allow for custom Lambda endpoints

## aws-lambda 3.3.0 17-Apr-2020

- Fix: when reusing the proxy based connection do not do the handshake again.
- revamped HTTP connect method that allows for connection reuse in scenarios
  with a proxy and/or ssl based connections

## aws-lambda 3.2.1 3-Mar-2020

- Maintenance release for CI purposes

## aws-lambda 3.2.0 11-Feb-2020

- Encrypt IAM access and secret keys when relevant

## aws-lambda 3.1.0 6-Jan-2020

- fix: reduce notice-level message to debug, to reduce log noise
- feat: added 3 regions; eu-north-1, me-south-1, eu-west-3

## aws-lambda 3.0.1 13-Nov-2019

- Remove the no-longer supported `run_on` field from plugin config schema

## aws-lambda 3.0.0 2-Oct-2019

- Renamed from `liamp` to `aws-lambda` to supersede the `aws-lambda` plugin
  from Kong core.
- Note that this version number is just to indicate it supersedes the
  previously build in one. The effective change from version 2.0 (build in)
  of the aws-lambda plugin to 3.0 (this) is the combined set of changes
  mentioned below for Liamp versions `0.1.0` and `0.2.0`.

## Liamp 0.2.0 23-Sep-2019

- chore: convert the plugin to the PDK and new DB (developed against Kong 1.x)

## Liamp 0.1.0

- feat: if no credentiuals are provided, the plugin will automatically fetch
  EC2 or ECS credentials and use the AWS IAM roles retrieved for accessing the
  Lambda.
- feat: new option `awsgateway_compatible` to make the serialized request
  compatible with the AWS gateway format, making the plugin a drop-in
  replacement
- feat: new option `skip_large_bodies` to enable really large bodies (that
  have been cached to disk) to also be sent to the Lambda. Use with care!
- feat: added the ability to connect to the Lambda through a proxy


kong/plugins/aws-lambda/schema.lua
local typedefs = require "kong.db.schema.typedefs"

local function keyring_enabled()
  local ok, enabled = pcall(function()
    return kong.configuration.keyring_enabled
  end)

  return ok and enabled or nil
end

-- symmetrically encrypt IAM access keys, if configured. this is available
-- in Kong Enterprise: https://docs.konghq.com/enterprise/1.3-x/db-encryption/
local ENCRYPTED = keyring_enabled()

return {
  name = "aws-lambda",
  fields = {
    { protocols = typedefs.protocols_http },
    { config = {
      type = "record",
      fields = {
        { timeout = {
          type = "number",
          required = true,
          default = 60000,
        } },
        { keepalive = {
          type = "number",
          required = true,
          default = 60000,
        } },
        { aws_key = {
          type = "string",
          encrypted = ENCRYPTED,
        } },
        { aws_secret = {
          type = "string",
          encrypted = ENCRYPTED,
        } },
        { aws_region = typedefs.host },
        { function_name = {
          type = "string",
          required = true,
        } },
        { aws_role_session_name = {
          type = "string",
          encrypted = ENCRYPTED,
        } },
        { aws_assume_role_arn = {
          type = "string",
          encrypted = ENCRYPTED,
        } },
        { qualifier = {
          type = "string",
        } },
        { invocation_type = {
          type = "string",
          required = true,
          default = "RequestResponse",
          one_of = { "RequestResponse", "Event", "DryRun" }
        } },
        { log_type = {
          type = "string",
          required = true,
          default = "Tail",
          one_of = { "Tail", "None" }
        } },
        { host = typedefs.host },
        { port = typedefs.port { default = 443 }, },
        { unhandled_status = {
          type = "integer",
          between = { 100, 999 },
        } },
        { forward_request_method = {
          type = "boolean",
          default = false,
        } },
        { forward_request_uri = {
          type = "boolean",
          default = false,
        } },
        { forward_request_headers = {
          type = "boolean",
          default = false,
        } },
        { forward_request_body = {
          type = "boolean",
          default = false,
        } },
        { is_proxy_integration = {
          type = "boolean",
          default = false,
        } },
        { awsgateway_compatible = {
          type = "boolean",
          default = false,
        } },
        { proxy_scheme = {
          type = "string",
          one_of = { "http", "https" }
        } },
        { proxy_url = typedefs.url },
        { skip_large_bodies = {
          type = "boolean",
          default = true,
        } },
      }
    },
  } },
  entity_checks = {
    { mutually_required = { "config.aws_key", "config.aws_secret" } },
    { mutually_required = { "config.proxy_scheme", "config.proxy_url" } },
    { only_one_of = { "config.aws_region", "config.host" } },
  }
}
